### DISCUSSION TITLE
Training on multi targets

### DISCUSSION THREAD

**Discussion by LtTuvok** [optuna/optuna#5150] [Q&A] [answered]
Hi

I have two targets. I want to optimize two sets of hyperparameters to get the minimum mae for each target independently. My optimization function returns a list of two mae values. This is also accepted in my study. The optimization runs but still when I ask for the trial results (study.get_trials()) I get just one value, where I expect the two mae values I return from the optimization function. 
As an example: I train a linear regression on two targets and the hyper_parameters are: 
`{
        'alpha_1': trial.suggest_float('alpha_1', 1e-2, 1e4, log=True),
        'alpha_2': trial.suggest_float('alpha_2', 1e-2, 1e4, log=True)
    },`

Here is an example of the study.get_trials():

FrozenTrial(number=0, state=TrialState.COMPLETE, values=[0.01758206833185925], datetime_start=datetime.datetime(2023, 12, 18, 13, 18, 15, 801977), datetime_complete=datetime.datetime(2023, 12, 18, 13, 18, 17, 505992), params={'alpha_1': 8.97750837360961, 'alpha_2': 0.012612236685184387}, user_attrs={}, system_attrs={}, intermediate_values={}, distributions={'alpha_1': FloatDistribution(high=10000.0, log=True, low=0.01, step=None), 'alpha_2': FloatDistribution(high=10000.0, log=True, low=0.01, step=None)}, trial_id=105, value=None),

I expect at values a second value in the list. Maybe I missed to set an option for my case? Can someone help me?

Regards

**Comment by nzw0301** [comment]
Could you share your objective function?

**Reply by LtTuvok** [reply]
Sure: 

```
func_opt_optuna_2 = lambda trial: func_opt_optuna(trial, dict_total, test_cycle)
study.optimize(func_opt_optuna_2, n_trials = n_optuna_trials)
```

with

```
def func_opt_optuna(trial, dict_total, test_cycle):

    df_data = pd.DataFrame(dict_total['data_info']['dict_data'])
    dict_para={}
    for name, value in dict_total['dict_optuna_paras'].items():
        dict_para[name] = eval(value)
    
    ls_features = dict_total['data_info']['ls_features']
    ls_targets = dict_total['data_info']['ls_targets']

    ls2_mae_dev_mean = []
    for dev_cycle in dict_total['test_splits'][test_cycle]['dev_splits']:

        ls_ind_train = dict_total['test_splits'][test_cycle]['dev_splits'][dev_cycle]['ls_ids_train_split']
        ls_ind_dev = dict_total['test_splits'][test_cycle]['dev_splits'][dev_cycle]['ls_ids_dev_split']
  
        df_X_train = df_data.loc[ls_ind_train,ls_features].copy()
        df_X_dev = df_data.loc[ls_ind_dev,ls_features].copy()
    
        df_y_train = df_data.loc[ls_ind_train,ls_targets].copy()
        df_y_dev = df_data.loc[ls_ind_dev,ls_targets].copy()

        ls_mae_dev = model_fit(dict_para, df_X_train, df_y_train, df_X_dev, df_y_dev, ls_targets)
        ls2_mae_dev_mean.append(ls_mae_dev)

    ls_mae_dev_mean = np.mean(ls2_mae_dev_mean)
    return(ls_mae_dev_mean)
```

and 

```
def model_fit(dict_para, df_X_train, df_y_train, df_X_dev, df_y_dev, ls_targets):
    model = linear_model.Ridge(alpha = np.array([dict_para['alpha_1'], dict_para['alpha_2']]))
    model.fit(df_X_train, df_y_train)
    ls_r2_dev, ls_mae_dev, _ = get_r2_mae(model, df_X_dev, ls_targets, df_y_dev)
    return(ls_mae_dev)
```

and

```
def get_r2_mae(model, df_X, ls_targets, df_y):
    np2_pred = model.predict(df_X)
    df_pred = pd.DataFrame(np2_pred, columns = [f'{target}_pred' for target in ls_targets])
    df_pred.index = df_y.index
    ls_r2 = r2_score(df_y, df_pred, multioutput='raw_values')
    ls_mae = mean_absolute_error(df_y, df_pred, multioutput='raw_values')

    return(ls_r2, ls_mae, df_pred)
```

**Reply by nzw0301** [reply]
Thanks. I think `func_opt_optuna` returns a single value by taking mean with `np.mean`, so 

> My optimization function returns a list of two mae values.

this understanding does not sound correct.

**Reply by LtTuvok** [reply]
Hi nzw0301
I checked and func_opt_optuna_2 does return a list with two entries, e.g. [0.1, 0.11]. Also the training runs as expected. I get good results. My problem is just that the optuna method get_trials() contains just one scalar:

[FrozenTrial(number=0, state=TrialState.COMPLETE, **values=[0.01758206833185925],** datetime_start...

Maybe this method is not correct?

**Reply by nzw0301** [reply]
Could you share reproducible code with us?

**Reply by LtTuvok** [reply, marked as answer]
Hi nzw0301
By creating the code I noticed, that indeed the np.mean returns the scalar, whereas I wanted np.mean(,0). Anyway, since it is too difficult to analyse the multiple xgboost hyperparameters optimized by optuna, I stick to the scalar and scale the targets before. Thank you!

**Reply by nzw0301** [reply]
Great. Thank you for your reply. I think multi-objective optimisation is a bit hard to select the best one due to the nature of https://en.wikipedia.org/wiki/Pareto_front, so a single-value approach as you said sounds a good first step. 

best regards,
